text/0003-placeholders-in-conditions.md

```diff
@@ -0,0 +1,145 @@
+
+This RFC was motivated by a particular customer use-case, i.e., [#81](https://github.com/cedar-policy/cedar/issues/81).
+
+# Detailed design
+
+This RFC does not propose any change to what syntax is valid in the policy scope. For instance, the following remains _invalid_:
+```
+permit(
+  principal,
+  action,
+  ?resource in resource
+);
+```
+The only valid uses of `?resource` in the policy scope are `resource == ?resource` and `resource in ?resource`, and this RFC is not proposing to change that. This RFC also disallows the use of `?principal` and `?resource` in the condition if they don't _also_ appear in the scope. 
+
+Both restrictions are in place because these two placeholder variables have special meaning for indexing policies in a policy store. In particular, looking _only_ at the linking information (i.e., to what `?principal` and/or `?resource` are bound) is sufficient to precisely select linked policies for particular requests. If the variables could only be bound in the condition or in other positions in the scope, and used in arbitrary contexts, it would lead to selecting superfluous policies. 
+
+Given these restrictions, the above policy would be impossible to write without support for placeholder variables beyond `?resource` and `?principal`. With their support, we can write: 
+```
+template (?lowerbound)
+permit(principal, action, resource)
+when { ?lowerbound in resource };
+```
+Here, the `template (?lowerbound)` part _introduces_ a fresh placeholder variable which can be used legally within the policy template that follows. Eliding the `template (?lowerbound)` prefix would be illegal, signaling an unbound variable. 
+
+Multiple variables can be introduced together, separated by commas, e.g.,
+```
+template (?user, ?resourcebound)
+permit(principal, action, resource)
+when { 
+     ?resourcebound in resource 
+  && principal in ?user.delegates
+};
+```
+
+Policy writers can choose to include `?principal` and/or `?resource` in the list of template variables if they like, but doing so is not required. These two variables retain their special status.
+
+If a variable is included in the list that does not appear in the actual template, we can issue a warning.
+
+All introduced placeholder variables presented so far are assumed to be entities (just like `?principal` and `?resource`). The particular _type_ of the entity is not given, for simplicity. Eliding the type poses a problem for validating templates on their own; we would have to wait to validate a template each time it is linked. However, link-time validation makes it straightforward to allow placeholder variables to have any type (`string`, `bool`, etc.), not just entity types. 
+
+Link-time validation has the benefit that it is lower overhead to write the templates, and potentially more flexible, since the particulars of the types don't need to be spelled out. New types added after the template is created that work with the template don't require changing the template.
```

Comment:
To be clear, this RFC is proposing that we typecheck templates only when linking?

Reply:
Yes. So we don't typecheck templates, just template instances. Is that a change from what we do now?

Reply:
Yes. We currently typecheck the template and then when checking a linked template we only need to see that linking won't cause the policy to have type `False`.

Reply:
I see. Then that's a breaking change. Does the change concern you?

Reply:
I don't think any existing templates/instantiations will fail to validate after this change: templates typechecking is obviously more permissive, and then typechecking the instantiations (for existing templates with the placeholders in the head) should not be able to cause any errors.

Reply:
Note that we could still do creation-time validation when templates involve only `?principal` and `?resource`. I just updated the RFC to make this point.

Reply:
We could, but I'm not convinced that we need to to maintain backwards compatibility.